nomt/src/cursor.rs

```diff
@@ -251,6 +414,77 @@ impl PageCacheCursor {
         }
     }
 
+    /// Attempt to compact this node with its sibling. There are four possible outcomes.
+    ///
+    /// 1. If both this and the sibling are terminators, this moves the cursor up one position
+    ///    and replaces the parent with a terminator.
+    /// 2. If one of this and the sibling is a leaf, and the other is a terminator, this deletes
+    ///    the leaf, moves the cursor up one position, and replaces the parent with the deleted
+    ///    leaf.
+    /// 3. If either or both is an internal node, this moves the cursor up one position and
+    ///    return an internal node data structure comprised of this and this sibling.
+    /// 4. This is the root - return.
+    pub fn compact_up(&mut self) -> Option<trie::InternalData> {
+        if self.depth == 0 {
+            return None;
+        }
+
+        let node = self.node();
+        let sibling = self.peek_sibling();
+
+        let this_bit_idx = self.depth as usize - 1;
+        // unwrap: depth != 0 above
+        let bit = *self.path.view_bits::<Msb0>().get(this_bit_idx).unwrap();
+
+        match (NodeKind::of(&node), NodeKind::of(&sibling)) {
+            (NodeKind::Terminator, NodeKind::Terminator) => {
+                // compact terminators.
+                self.up(1);
+                self.place_non_leaf(trie::TERMINATOR);
+                None
+            }
+            (NodeKind::Leaf, NodeKind::Terminator) => {
+                // compact: clear this node, move leaf up.
+
+                let prev = self.read_leaf_children();
+                self.write_leaf_children(Default::default());
+                self.place_non_leaf(trie::TERMINATOR);
+                self.up(1);
+                self.place_leaf(node, prev.clone());
+
+                None
+            }
+            (NodeKind::Terminator, NodeKind::Leaf) => {
```

Comment:
Can't we merge this arm with the previous one, by stating the both patterns in one arm?

Reply:
The `self.sibling()` call is important